### DISCUSSION TITLE
Consider to use CoW instead of hard links on ReFS (and other modern FS)

### DISCUSSION THREAD

**Discussion by huoyaoyuan** [ppy/osu#27014] [Ideas]
## Background

Copy on Write (CoW) is a filesystem feature that allows different files to share the same storage blocks. It provides similar benefits to hard links, but is safer that if one file is written, a copy will be allocated and other files will remain unchanged.

CoW is only supported on newer filesystems. It's not supported by the most widely used NTFS. The next generation filesystem of Windows, ReFS does support this feature. See https://en.wikipedia.org/wiki/Comparison_of_file_systems
*Notice: ReFS is still in evolving. Insider versions of Windows may upgrade the version of ReFS drive. DO NOT use it unless you are very advanced user.*

Historically, CoW operation requires different syscall on it. However, the Windows team plans to make the copy syscall defaults to CoW in next release of Windows (https://github.com/dotnet/runtime/issues/86681#issuecomment-1781550931). This means that the default copy operation will just work.

## Proposed Work

I'm proposing change because there's no actual work need to be performed at our side. Instead, we need to add some indicator that **hard link should not be used because copying is just better**. The condition should be Windows version number > 25xxx and the drive is ReFS.

## Other Platforms

On older versions of Windows, the manual operation can be done via the Microsoft.CopyOnWrite package. But I don't think it necessary give the adoption rate of ReFS.
Similar operation may also be done manually on ZFS/APFS etc.

**Comment by Rekunan** [comment]
I feel like it'd be better to only have to manage one form of "linkage," especially considering the lack of older support.

But also, wouldn't this be doing something different from what was intended with hard links? Iirc, the only goal was to create a way so that you could use the preexisting data from stable in lazer without having to take up double the disk space.

**Reply by huoyaoyuan** [reply]
The difference is whether edit of one version will affect another version.

If not explicit required, it should be better to not make edits visible. It may also break implications like file hash in Lazer.

**Reply by Rekunan** [reply]
Actually I think that's already a thing.
https://osu.ppy.sh/wiki/en/Client/Release_stream/Lazer/File_storage#via-hard-links
> Updating or editing files causes the files to become separate, and will therefore also not affect the other installation, as long as osu! is used to make the changes.

**Reply by huoyaoyuan** [reply]
> as long as osu! is used to make the changes

This is a bit strong. CoW can isolate from *any* modification.

**Reply by Rekunan** [reply]
Well I think considering they want to protect osu from outside file changes (https://github.com/ppy/osu/wiki/User-file-storage), that sounds like a bad idea.

**Reply by huoyaoyuan** [reply]
> I feel like it'd be better to only have to manage one form of "linkage," especially considering the lack of older support.

I feel hard link as a workaround for the absense of CoW.

The reason I'm proposing this for ReFS is the upcoming change of Windows. It will be "smart copy" that optimises disk usage. At the application's view, it's just a simple copy operation. No more action is required, so there's actually no new form introduced.
